# Assignment 2: Linear Models and Validation Metrics (30 marks total)
### Due: October 10 at 11:59pm

### Name: 

### In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.

## Part 1: Classification (14.5 marks total)

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:

### Step 0: Import Libraries

In [161]:
import numpy as np
import pandas as pd



### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [162]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y

from yellowbrick.datasets import load_spam
X, y = load_spam()
print("X size =",X.size, "y size= ", y.size, "X types: ", type(X), "y types: ", type(y),y.dtypes)
X.head()

X size = 262200 y size=  4600 X types:  <class 'pandas.core.frame.DataFrame'> y types:  <class 'pandas.core.series.Series'> int64


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.21               0.28           0.50           0.0   
1            0.06               0.00           0.71           0.0   
2            0.00               0.00           0.00           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.14            0.28              0.21                0.07   
1           1.23            0.19              0.19                0.12   
2           0.63            0.00              0.31                0.63   
3           0.63            0.00              0.31                0.63   
4           1.85            0.00              0.00                1.85   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0             0.00            0.94  ...                   0.0         0.00   
1             0.64            0.25  ...                   0.0         0.01   
2             0.31            0.63  ...                   0.0         0.00   
3             0.31            0.63  ...                   0.0         0.00   
4             0.00            0.00  ...                   0.0         0.00   

   char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0        0.132          0.0        0.372        0.180        0.048   
1        0.143          0.0        0.276        0.184        0.010   
2        0.137          0.0        0.137        0.000        0.000   
3        0.135          0.0        0.135        0.000        0.000   
4        0.223          0.0        0.000        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       5.114                         101   
1                       9.821                         485   
2                       3.537                          40   
3                       3.537                          40   
4                       3.000                          15   

   capital_run_length_total  
0                      1028  
1                      2259  
2                       191  
3                       191  
4                        54  

[5 rows x 57 columns]

### Step 2: Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [163]:
# TO DO: Check if there are any missing values and fill them in if necessary
X.isnull().values.any()

False

For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [164]:
# TO DO: Create X_small and y_small 
from sklearn.model_selection import train_test_split
X_small, X_test, y_small, y_test = train_test_split(X, y, 
                                                    test_size=0.05,
                                                    stratify=y,
                                                    random_state=0)


### Step 3: Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets: 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

### Step 4: Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

### Step 5: Visualize Results (4 marks)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [165]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_validate
#step 3
model = LogisticRegression(max_iter=2000)
model.fit(X_small, y_small)
model.fit(X,y)
model.fit(X.iloc[:,:2],y)
##step 4
result={}
datasets=[ (X, y, 'X_and_y'),(X_small, y_small, 'small data'), (X.iloc[:, :2], y, 'First_Two_Columns')]
for X_data, y_data, column in datasets:
    scores = cross_validate(model, X_data, y_data, cv=5, 
                        scoring='accuracy',
                       return_train_score=True)
    train_score=scores['train_score'].mean()
    test_score=scores['test_score'].mean()
    result[column]={'training_score':train_score,'validation_score':test_score, 'Difference':train_score-test_score}
#step 5
results=pd.DataFrame(data=result)
results.head()

X_and_y  small data  First_Two_Columns
training_score    0.930978    0.931979           0.616250
validation_score  0.912826    0.926773           0.609783
Difference        0.018152    0.005206           0.006467

### Questions (4 marks)
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

*YOUR ANSWERS HERE*

1. When only 5% of the data is used in the model, it does not show a significant change in the validation score compared to the model fitted to the entire dataset. The trainning score remained 0.93, but the validation score increased by 0.01 when the small data was used. The difference between the training score and the validation score is 0.005 for the small data model, and 0.018 for the full data model, which means there is more bias in the small data model. The model fitted to the first two columns of X showed a significant decrease in performace compared to the other two datasets. The data used was to less for the model to learn from, and if there was noise, that may be cause the performance to drop as well.

2. Since this model is based on emails being marked as spam, a false positive represents an email that is not spam being sent to the spam folder and a false negaive represents an email that is spam not being marked as such. In this case, false positives may be worse because they can cause a person to miss an important email.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code? 
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*
I mostly used the Linear Example-filled.ipynb file provided to complete this section of the assignment. The links to the other sources used are provided below. I completed the steps in order, and the most challenging part for me were steps 4 and 5. I was attempting to create the dataframe of the results from an array and was not able to and decided to use a dictionary with key-value pairs instead. 

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html

https://datascience.stackexchange.com/questions/14152/does-dataset-training-and-test-size-affect-algorithm#:~:text=A%20larger%20training%20set%20decreases%20the%20score%20because,an%20increase%20in%20the%20model%27s%20ability%20to%20generalise.

https://www.geeksforgeeks.org/python-creating-dataframe-from-dict-of-narray-lists/



## Part 2: Regression (10.5 marks total)

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [173]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y
from yellowbrick.datasets import load_concrete
X, y = load_concrete()
print("X size =",X.size, "y size= ", y.size, "X types: ", type(X), "y types: ", type(y),y.dtypes)
X.head()

X size = 8240 y size=  1030 X types:  <class 'pandas.core.frame.DataFrame'> y types:  <class 'pandas.core.series.Series'> float64


cement   slag  ash  water  splast  coarse   fine  age
0   540.0    0.0  0.0  162.0     2.5  1040.0  676.0   28
1   540.0    0.0  0.0  162.0     2.5  1055.0  676.0   28
2   332.5  142.5  0.0  228.0     0.0   932.0  594.0  270
3   332.5  142.5  0.0  228.0     0.0   932.0  594.0  365
4   198.6  132.4  0.0  192.0     0.0   978.4  825.5  360

In [167]:
# y=pd.to_numeric(y, errors='coerce').astype('int64')
# y.dtypes


### Step 2: Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [174]:
# TO DO: Check if there are any missing values and fill them in if necessary
X.isnull().any()


cement    False
slag      False
ash       False
water     False
splast    False
coarse    False
fine      False
age       False
dtype: bool

### Step 3: Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with `X` and `y`

In [175]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0
from sklearn.linear_model import LinearRegression
# from sklearn import preprocessing
# scaler = preprocessing.StandardScaler().fit(X)
# x_scaled=scaler.transform(X)

X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
lr = LinearRegression().fit(X_train, y_train)

### Step 4: Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [170]:
# TO DO: ADD YOUR CODE HERE
from sklearn.metrics import mean_squared_error, r2_score
y_tpred=lr.predict(X_train)
y_vpred=lr.predict(X_val)

r2 = r2_score(y_train, y_tpred)
mse=mean_squared_error(y_train,y_tpred)

r2_v = r2_score(y_val, y_vpred)
mse_v=mean_squared_error(y_val,y_vpred)
result={}
datasets=[ (y_train, y_tpred, 'Training accuracy'),(y_val, y_vpred, 'Validation accuracy')]
for y, y_pred, column in datasets:
    r2 = r2_score(y, y_pred)
    mse=mean_squared_error(y,y_pred)
    result[column]={'mse':mse,'r2':r2}

### Step 5: Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [171]:
# TO DO: ADD YOUR CODE HERE
results=pd.DataFrame(data=result)
print(results)

     Training accuracy  Validation accuracy
mse         111.358439            95.904136
r2            0.610823             0.623414


### Questions (2 marks)
1. Did using a linear model produce good results for this dataset? Why or why not?

Based on the R2 values, the model produced fairly good results. With accuracies of approximately 61% for the training data and 62% for the validation data. The performance was better for the validation data, however it was not that far off from the performance for the training data. The mean square error is 111 for the training data and 96 for validation, and a greater value generally indicates that the model was not a good fit. This could however also be caused by numerical discrepencies, such as when the head of the data was printed, the column containing the age increased from 28 at index 0 to 360 at index 4. Although, in this case it might be that this model was not a good fit because the first sentence on the website provided in step 1 says that concrete compressive strength is a highly nonlinear funciton.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*
I completed the steps in order for this part as well. I completed part 3 using LogisticRegression, and noticed that I was supposed to do LinearRegression. The links provided to the sources used are in the order they were accessed. While I was doing logistic regression, I converted y from float to int type since classification can only be used when the target is categorical. My model was not converging with 2000 iterations for which I used the pre-processing feature. For the linear regression model, I used Linear Regression Python file provided on D2L under examples.
links to sources:
https://www.statology.org/valueerror-unknown-label-type-continuous/
https://stackoverflow.com/questions/65463794/valueerror-unknown-label-type-continuous-in-decisiontreeclassifier
https://stackoverflow.com/questions/25952790/convert-pandas-series-from-dtype-object-to-float-and-errors-to-nans
https://scikit-learn.org/stable/modules/preprocessing.html
https://www.statology.org/r-squared-in-python/



## Part 3: Observations/Interpretation (3 marks)

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


During part 1, I printed the head of both datasets, and both contained many 0's for a lot of the features. I also printed the size of the dataset and both datsets were large. In class we discussed that linear models fit well to large datasets and are able to work well with sparse data. For part 1, the accuracy was 91% when the full data was used and for part 2 the accuracy was 61%. This shows that the linear model fit the data well. 

We also discussed that linear models perform well when there are more features compared to the samples. In part 1, when only the first 2 columns were used, the number of features were significantly reduced comapred to the samples. This resulted in a significant drop in the accuracy compared to the models fitted to the full data and 5% of the data.

## Part 4: Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*
I liked that the example codes provided are really helpful for a lot of the difficult steps. I found it challenging to interpret the results, especially for the second part.

## Part 5: Bonus Question (4 marks)

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [202]:
# TO DO: ADD YOUR CODE HERE
from sklearn.linear_model import Ridge
from yellowbrick.datasets import load_concrete
X, y = load_concrete()
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
list=[i/1000 for i in range(0,100001,100)]
for l in list:
    print(l)
    ridge10 = Ridge(alpha=l).fit(X_train, y_train)
    print("Ridge Training score: {:.2f}".format(ridge10.score(X_train, y_train)))
    print("Ridge Validation score: {:.2f}".format(ridge10.score(X_val, y_val)))


0.0
Ridge Training score: 0.61
Ridge Validation score: 0.62
0.1
Ridge Training score: 0.61
Ridge Validation score: 0.62
0.2
Ridge Training score: 0.61
Ridge Validation score: 0.62
0.3
Ridge Training score: 0.61
Ridge Validation score: 0.62
0.4
Ridge Training score: 0.61
Ridge Validation score: 0.62
0.5
Ridge Training score: 0.61
Ridge Validation score: 0.62
0.6
Ridge Training score: 0.61
Ridge Validation score: 0.62
0.7
Ridge Training score: 0.61
Ridge Validation score: 0.62
0.8
Ridge Training score: 0.61
Ridge Validation score: 0.62
0.9
Ridge Training score: 0.61
Ridge Validation score: 0.62
1.0
Ridge Training score: 0.61
Ridge Validation score: 0.62
1.1
Ridge Training score: 0.61
Ridge Validation score: 0.62
1.2
Ridge Training score: 0.61
Ridge Validation score: 0.62
1.3
Ridge Training score: 0.61
Ridge Validation score: 0.62
1.4
Ridge Training score: 0.61
Ridge Validation score: 0.62
1.5
Ridge Training score: 0.61
Ridge Validation score: 0.62
1.6
Ridge Training score: 0.61
Ridge Val

In [198]:
from sklearn.linear_model import Lasso
from yellowbrick.datasets import load_concrete
X, y = load_concrete()
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
list=[i for i in range(101)]
for l in list:
    print(l)
    lasso001 = Lasso(alpha=l, max_iter=100000).fit(X_train, y_train)
    print("Training set score: {:.2f}".format(lasso001.score(X_train, y_train)))
    print("Validation set score: {:.2f}".format(lasso001.score(X_val, y_val)))
    print("Number of features used:", np.sum(lasso001.coef_ != 0))

0


c:\Users\tejpr\anaconda3\envs\ensf-ml\lib\site-packages\sklearn\base.py:1152: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  return fit_method(estimator, *args, **kwargs)
c:\Users\tejpr\anaconda3\envs\ensf-ml\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  model = cd_fast.enet_coordinate_descent(
c:\Users\tejpr\anaconda3\envs\ensf-ml\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.298e+04, tolerance: 2.209e+01 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_model.Ridge/RidgeCV in

Training set score: 0.61
Validation set score: 0.62
Number of features used: 8
1
Training set score: 0.61
Validation set score: 0.62
Number of features used: 8
2
Training set score: 0.61
Validation set score: 0.63
Number of features used: 8
3
Training set score: 0.61
Validation set score: 0.63
Number of features used: 8
4
Training set score: 0.61
Validation set score: 0.63
Number of features used: 8
5
Training set score: 0.61
Validation set score: 0.63
Number of features used: 7
6
Training set score: 0.61
Validation set score: 0.63
Number of features used: 7
7
Training set score: 0.61
Validation set score: 0.63
Number of features used: 7
8
Training set score: 0.61
Validation set score: 0.63
Number of features used: 7
9
Training set score: 0.60
Validation set score: 0.63
Number of features used: 7
10
Training set score: 0.60
Validation set score: 0.63
Number of features used: 6
11
Training set score: 0.60
Validation set score: 0.63
Number of features used: 6
12
Training set score: 0.60


*ANSWER HERE*
For the ridge regression, I attempted to test the accuracy for most of the values from 0.001 to 100 and it did not seem to change at all. The traning score remained at 0.61 and the validation at 0.62 which is the result I got from the linear regression.

The accuracy of lasso regression for alpha values less than 1 and up to 100 showed a slight increase in validation score from the part 2 results. In part 2 it was 0.62 and the best using lasso regression is 0.63 while the training score remained at 0.61.